presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java

```diff
@@ -624,6 +625,9 @@ public PlanOptimizers(
 
         // DO NOT add optimizers that change the plan shape (computations) after this point
 
+        // Merge join - optimizations to use merge join when data properties meet requirements
+        builder.add(new MergeJoinOptimizer(metadata, sqlParser));
```

Comment:
The comment above seems to indicate we shouldn't change the plan shape.  But I believe we are, since we're eliminating exchanges?  Should we consider moving this before `AddExchanges`, then performing a minor alteration on `AddExchanges` to skip over merge join nodes?

Reply:
Good point. Does it work if we add this rule right after the AddLocalExchanges rule? In that case, we don't need to update the AddLocalExchanges rule.

Reply:
I strongly agree with @tdcmeehan.  I think the structure should be as follows

1. Add rule to turn hash join -> merge join
2. Change AddLocalExchanges not to add a localExchange for merge join.

Reply:
I agree. This structure is cleaner.

Reply:
I remember we used to say the plan is completely logical before AddExchange and AddLocalExchange, and we don't access data layout properties until those two optimization rule. 
After that, the plan becomes physical.

However MergeJoinOptimizer has to access all the data layout and if we move it before AddLocalExchanges, it changes all the assumptions above

another example is `RuntimeReorderJoinSides`, it's also placed after `AddLocalExchange`

thoughts @tdcmeehan @rschlussel